### ISSUE TITLE
building py-yt now fails with cannot depend twice error

### ISSUE THREAD

**Issue by krafczyk** [LLNL/spack#2728]
running `spack spec py-yt` now gives:

```
spack spec py-yt
Input spec
--------------------------------
py-yt

Normalized
--------------------------------
py-yt
    ^py-cython
        ^python@2.7:2.8,3.4:
            ^bzip2
            ^ncurses
            ^openssl
                ^zlib@1.0.4:
            ^readline
            ^sqlite
    ^py-ipython
        ^py-pygments
            ^py-setuptools@17.1:
    ^py-matplotlib
        ^freetype@2.3:
            ^libpng@1.2:
            ^pkg-config@0.9.0:
        ^py-cycler@0.9:
            ^py-six@1.9.0:
        ^py-dateutil@1.1:
        ^py-mock
            ^py-pbr
        ^py-nose
        ^py-numpy@1.6:
        ^py-pyparsing
        ^py-pytz
        ^qhull@2012.1:
            ^cmake@2.6:
    ^py-sympy

Concretized
--------------------------------
==> Error: Cannot depend on 'py-pillow^binutils@2.26:^bison^bzip2@1.0.6%gcc@6.2.1 arch=linux-arch-x86_64^flex^gettext@0.19:^help2man^m4^ncurses@6.0%gcc@6.2.1 arch=linux-arch-x86_64^openssl@1.0.2j%gcc@6.2.1 arch=linux-arch-x86_64^py-setuptools@25.2.0%gcc@6.2.1 arch=linux-arch-x86_64^python@2.7.13%gcc@6.2.1~tk+ucs4 arch=linux-arch-x86_64^readline@6.3%gcc@6.2.1 arch=linux-arch-x86_64^sqlite@3.8.5%gcc@6.2.1 arch=linux-arch-x86_64^zlib@1.2.10%gcc@6.2.1+pic arch=linux-arch-x86_64' twice
```

a quick git bisect reveals that this behaviour is introduced at commit 5fbab1f4b553b9ec550dfd0eef3078e7c87720aa

**Comment by krafczyk** [comment]
@tgamblin, I believe 5fbab1f is your commit, any idea what's going wrong?

**Comment by adamjstewart** [comment]
Might be a problem with virtual deptypes? `py-pillow` provides `pil`.

**Comment by krafczyk** [comment]
I created the following test package to see what causes this problem.

```
from spack import *

class PyPilTest(Package):
    """Description"""

    homepage = "http://www.example.com"
    url      = "http://www.example.com/py-pil-test-1.0.tar.gz"

    version('1.0', '0123456789abcdef0123456789abcdef')
    
    variant("astropy", default=True, description="enable astropy support")

    depends_on("py-astropy", type="nolink", when="+astropy")
    depends_on("py-matplotlib", type="nolink")
    depends_on("python @2.7:2.999,3.4:")
    
    def install(self, spec, prefix):
        pass
```

It is based on `py-yt`. Interestingly if I remove the variant `astropy` and remove the `when="+astropy"` from the `py-astropy` dependency, the error goes away.

Also, if I add an explicit dependency on `py-pillow` the error goes away.

For the time being I'm going to use the explicit `py-pillow` dependency as a workaround.

**Comment by eschnett** [comment]
I encountered a similar issue. I had a package depend on `blas`, `lapack`, and `openblas`. Removing the `blas` and `lapack` dependencies made the error go away.

**Comment by adamjstewart** [comment]
Oh, I completely forgot about this issue. @eschnett You are exactly right. See #2807 for a more detailed, generic explanation of the problem. I fixed the `py-yt` spec issue in #2778 by replacing `py-pillow` with `pil` (see [here](https://github.com/LLNL/spack/pull/2778/files)). I'm going to close this issue, as the `py-yt` issue is now solved. We can continue the more generic discussion of whether or not this should crash concretization in #2807.
